src/steamship/agents/schema/chathistory.py

```diff
@@ -272,3 +278,120 @@ def clear(self):
             self.embedding_index.reset()
 
         self.refresh()
+
+    def append_agent_message(
+        self,
+        text: str = None,
+        tags: List[Tag] = None,
+        content: Union[str, bytes] = None,
+        url: Optional[str] = None,
+        mime_type: Optional[MimeTypes] = None,
+    ) -> Block:
+        """Append a new block to this with status update messages from the Agent."""
+        return self.append_message_with_role(text, RoleTag.AGENT, tags, content, url, mime_type)
+
+    def append_tool_message(
+        self,
+        text: str = None,
+        tags: List[Tag] = None,
+        content: Union[str, bytes] = None,
+        url: Optional[str] = None,
+        mime_type: Optional[MimeTypes] = None,
+    ) -> Block:
+        """Append a new block to this with status update messages from the Agent."""
+        return self.append_message_with_role(text, RoleTag.TOOL, tags, content, url, mime_type)
+
+    def append_llm_message(
+        self,
+        text: str = None,
+        tags: List[Tag] = None,
+        content: Union[str, bytes] = None,
+        url: Optional[str] = None,
+        mime_type: Optional[MimeTypes] = None,
+    ) -> Block:
+        """Append a new block to this with status update messages from the Agent."""
+        return self.append_message_with_role(text, RoleTag.LLM, tags, content, url, mime_type)
+
+
+class ChatHistoryLoggingHandler(StreamHandler):
+    """Logs messages emitted by Agents and Tools into a ChatHistory file.
+
+    This is a basic mechanism for streaming status messages alongside generated content.
+    """
+
+    chat_history: ChatHistory
+    log_level: any
+    streaming_opts: StreamingOpts
+
+    def __init__(
+        self,
+        chat_history: ChatHistory,
+        log_level: any = logging.INFO,
+        streaming_opts: Optional[StreamingOpts] = None,
+    ):
+        StreamHandler.__init__(self)
+        formatter = FluentRecordFormatter(LOGGING_FORMAT, fill_missing_fmt_key=True)
+        self.setFormatter(formatter)
+        self.chat_history = chat_history
+        self.log_level = log_level
+        if streaming_opts is not None:
+            self.streaming_opts = streaming_opts
+        else:
+            self.streaming_opts = StreamingOpts()
+
+    def emit(self, record):
+        if record.levelno < self.log_level:
+            # don't bother doing anything if level is below logging level
+            return
+
+        message_dict = cast(dict, self.format(record))
```

Comment:
[not high pri; feel free to ignore] 

below feels like an opportunity to collapse into one block of code below since so much is repeated.

Reply:
refactored a bit. the slight diffs between the various modes makes a full collapse uglier than a small bit of near repetition (imho), but the refactor hopefully makes things clearer/more readable.